include/misc/__assert.h

```diff
@@ -67,6 +67,11 @@
 #endif
 #endif
 
+#ifdef CONFIG_FORCE_NO_ASSERT
+#undef __ASSERT_ON
```

Comment:
Nasty. I can only imagine how much time was sunk tracking this down.
Suggest an upstream report to gcc team if you haven't already.